### ISSUE TITLE
improve edge cases with the way get_state is called in simulator samples

### ISSUE THREAD

**Issue by ForrestTrepte** [microsoft/microsoft-bonsai-api#79]
Three related issues:
1. Some simulators retrieve state from data structures that have not been initialized until an episode has begun. In this case, it is inconvenient to have to implement a get_state method that is safe to call right after the sim object has been created before it has been initialized in a config. Sim developers shouldn't be required to handle the zen-like edge case of "What is the state of the environment when the environment doesn't yet exist?"
2. Although get_state shouldn't, in theory, have side effects sometimes this can happen. For example, if get_state is performing bookkeeping operations such as counting iterations or taking deltas to compute velocity from a previous state value. Although, ideally that should be done in the step function instead of the get_state function, it would be nice to only call get_state once per iteration in order to avoid such issues.
3. The behavior of the main() function should match the behavior of the test_policy() function. Currently they differ with respect to the previous two points.

To sum up, in samples such as [https://github.com/microsoft/microsoft-bonsai-api/blob/main/Python/samples/cartpole/main.py](Python Cartpole), for both the test path and the main path:
* get_state should only be called when an episode is active. (EpisodeStart has been called and EpisodeStop hasn't been called to end that episode.)
* get_state should only be called once per iteration
